### DISCUSSION TITLE
General improvements for streaming over the web

### DISCUSSION THREAD

**Discussion by dgcampea** [jcorporation/myMPD#732] [Ideas]
## Feature request: _MediaSession_ interface (allow control using Media Keys)

MediaSession allows controlling mympd with keyboard media keys or through a interface in the notification/lockscreen. This would improve mympd UX on a smartphone.
https://developer.mozilla.org/en-US/docs/Web/API/MediaSession

## Improvement: Improve control responsiveness when using _Local playback_
When pressing `Next`, it may occasionally take several seconds for it to take effect due to audio buffering (most evident when streaming over the internet).
Disconnecting and reconnecting seems to flush the stale buffered audio data and improve the reactivity of the button.
A possible improvement here would be making the 'Next' button perform some flushing of the audio data or simulating a 'reconnect' of the _Local playback_.

## Improvement: Option to begin _Local playback_ automatically on connect
Since I'm primarily using it behind a reverse proxy, I'm thinking that this could be implemented as a (request) header that is set by nginx (maybe `X-MYMPD-STREAM=1`?) to indicate whether mympd should automatically start _Local playback_ without user interaction.
The idea is to allow the use of nginx to perform some magic with user-agents, etc. to set this behavior.

### Security: Implement `Permissions-Policy` (previously known as `Feature-Policy`) header
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy

### Security: Remove `X-XSS-Protection` header
CSP already provides this protection and most browsers do not support this header per https://caniuse.com/mdn-http_headers_x-xss-protection.
According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection, it shouldn't be used as it may actually reduce the protection against XSS.

### Security: Replace obsolete `X-Frame-Options` with CSP policy
Can be replaced with `frame-ancestors 'none'`.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors

## Performance improvement: "Slow server times" when fetching covers
The album covers seem to be fetched sequentially and generally take more than 1 second to load.
Using the Network tool under developer tools it seems that most image requests take aprox. 1.5s and are indicated with a turtle icon ("Slow server response time (...). The recommended limit is 500ms).
Raising or loweing the binary limit doesn't seem to improve this. (mpd host is `localhost`, a slow link problem seems unlikely here)
Maybe fetching the covers using multiple threads would improve the situation?

**Comment by jcorporation** [comment]
# Feature request: MediaSession interface (allow control using Media Keys)

I played in the past with the MedieSession interface, but it has a restriction: There is no notification for audio from the Web Audio API unless it is played back via an audio element (https://developer.chrome.com/blog/media-notifications/). As I read this, it is not possible to control MPD playback with this interface.

**Reply by dgcampea** [reply]
What about the 'Local playback' case?

**Reply by jcorporation** [reply]
I think this is not good idea. Most users would expect to control mpd with this controls.

However this api is already implemented in myMPD and could be enabled in settings.

**Reply by jcorporation** [reply]
I just checked the media session implementation : it works with local playback.

**Reply by dgcampea** [reply]
Does 'Previous' and 'Next' work? I've tested it and the only thing that seems to work is 'Play/Pause'

**Reply by jcorporation** [reply]
On my iPhone I have no prev/next buttons. Play/Pause works and the title with albumart is displayed.
But with Chrome on my laptop the prev/next and the seek buttons are working.

**Reply by dgcampea** [reply]
Fedora 35 + GNOME:
 - Firefox (Xorg?), both software/hardware Play/Pause work but Prev/Next don't. (it shows up in Notifications and Lockscreen).
 - Chromium (XWayland), shows up in software with only Play/Pause enabled. Clicking on it doesn't do anything though.
 - Firefox (Wayland), neither work. I suspect this might be due to some flatpak permission issues that need to be investigated further.

Android + Firefox:
 - Play/Pause shown in notification.  Prev/Next not present.

The Linux case is the most erratic so far, sometimes the Play/Pause works other times it doesn't but this might be some browser/DE quirk.

**Reply by jcorporation** [reply]
My test was with chromium with Ubunut 22.04. It seems the browser implementations are not as major as it should be.

**Comment by jcorporation** [comment]
# Improvement: Improve control responsiveness when using Local playback

Buffering is done by the browser and different browsers behave a bit different. The stop button for local playback handles this and resets the audio element to workaround this behavior, flushing the buffer seems not possible. Adding this workaround to the next, prev, stop and pause buttons should fix this buffering issue.

I opened issue https://github.com/jcorporation/myMPD/issues/734 for this.

**Reply by jcorporation** [reply]
This is now implemented in the v9.3.0 branch. It works generally, but needs further testing.

**Comment by jcorporation** [comment]
# Improvement: Option to begin Local playback automatically on connect

Local playback must be started from the browser side, not from the server side. Most browsers do not allow autoplay in the default settings. It must be enabled by the user. Older versions of myMPD had this feature, but I removed it because of browsers default behavior. 

I opened issue https://github.com/jcorporation/myMPD/issues/734 for this.

**Reply by dgcampea** [reply]
I don't mind explicitly whitelisting autoplay for the domain I'm using as I'm the only person using it but since I'm using the same domain both for streaming on "remote devices" and on the host running mpd+mympd as a simple frontend, I'd prefer it to be a feature toggled through a request header (to avoid a big confusion of multiple mympd instances).

**Reply by jcorporation** [reply]
This would be an option that must be explicitly enabled for a myMPD instance. The browser setting is for the specific site or global.

**Reply by dgcampea** [reply]
I'm using a reverse proxy setup (nginx) to connect to the mympd instance and have `servername` set to a globally resolvable domain that points to the GUA IPv6 address of the host. (no tinkering with NAT or routing tables required)
In this way, I can use \<mydomain\> to simultaneously:
* locally access mympd to use as frontend only
* stream to lan or wan devices

And keep the same (smart)playlists/jukebox/stickers instead of having disjoint sessions. This single \<mydomain\> can also be whitelisted on my devices as I'm the one managing it anyways.
Locally, it makes little sense to have mympd start a 'Local playback' session as mpd is already outputting to the pipewire output.
In my case, when accessing from remote devices, the next action that I do after authenticating (handled by reverse proxy) is to immediately start 'Local playback' which gets repetitive over time.
Since I'm already using a reverse proxy setup, the idea was to use its extra flexibility to have it perform some decisions based on user-agent or client source IP/subnets to intelligently automate this task.

**Reply by jcorporation** [reply]
Ok, I think I understand it now. An autoplay option stored in the browser local storage would be sufficient?

**Reply by dgcampea** [reply]
> Ok, I think I understand it now. An autoplay option stored in the browser local storage would be sufficient?

Sure, didn't think about it but it sounds much simpler than what I was thinking.

**Reply by jcorporation** [reply]
This is now implemented in the v9.3.0 branch. It works generally, but needs further testing.

**Reply by dgcampea** [reply]
Is the v9.3.0 branch considered unrelated to the devel branch? I usually use the devel container for testing recent changes.

**Reply by jcorporation** [reply]
Right, the v9.3.0 branch is separate. The devel branch is only for bugfixes (patch releases). Minor and major releases are developed in separate branches.

**Reply by jcorporation** [reply]
The v9.3.0 branch was merged into devel.

**Reply by dgcampea** [reply]
Given that mympd has scripting support, I was thinking if it was possible for scripts to read "cookies" or HTTP headers and automatically set some mympd settings based on their presence or values.

**Reply by jcorporation** [reply]
This is not possible. Scripts are executed in a separate thread that is detached and can only communicate over a async message queue with the other threads. But I am interested in what would you achieve. It’s worth a another discussion.

**Comment by jcorporation** [comment]
# Header improvements

> Security: Remove X-XSS-Protection header
> Security: Replace obsolete X-Frame-Options with CSP policy

Will be removed in the next release.

> Security: Implement Permissions-Policy (previously known as Feature-Policy) header

Indeed a interesting new feature. I will have a look into it. 

I opened issue #733  for this.

**Reply by jcorporation** [reply]
This is now implemented in the v9.3.0 branch. With this Permissions-Policy header autoplay works for me without allowing it in the browser.

**Comment by jcorporation** [comment]
# Performance improvement: "Slow server times" when fetching covers

Yes, the images are requested sequentially from MPD, because myMPD maintains only a single connection to MPD.

myMPD could also read the images directly without the usage of the MPD protocol. This is enabled if myMPD has access to the MPD music directory and should be much faster.

If MPD and myMPD runs on the same host you should use a local socket not "localhost" this should also be faster and myMPD could autodetect the MPD music directory.

**Reply by dgcampea** [reply]
It's not entirely trivial to get socket working if mympd is running in a rootless container.
The mympd container starts as root and then drops to mympd user, this makes it impossible to use `docker/podman run --group-add` on mympd user to grant privileges to read the mpd socket on the host system. To complicate things further, if mpd is also running as non-root, it defaults to `0600` socket permissions.
An ugly workaround I did was monkeying around with systemd:
```
$ cat ~/.config/systemd/user/mpd.service.d/override.conf
[Service]
# https://github.com/MusicPlayerDaemon/MPD/blob/c5bf7948ff09776b3196a59ebc632aa3e58e6a27/src/Listen.cxx#L102
# adjust socket permissions for mympd
ExecStartPost=/usr/bin/chmod +rwx %t/mpd
ExecStartPost=/usr/bin/chmod 666 %t/mpd/socket
```
And pass `-v %t/mpd:/run/mpd` (or $XDG_RUNTIME_DIR instead of %t) to the docker/podman run command.

Containers also result in 'autodetect music directory' to fail unless the Music directory happens to be mounted with the exact same path configuration, although this can be remedied by manually setting the music directory.
Regarding the images, they still seem to be fetched using mpd `readpicture` and `albumart` according to mympd logs and I didn't notice any improvements in the loading speed.

**Reply by jcorporation** [reply]
Containers complicates this thing a bit, but it should be possible. Most speed improvement would be achieved by the direct access to the music directory for myMPD.

Things to test:
- Mount the music directory as volume and set the music directory manually in the myMPD connection dialog
- Start the myMPD container not as root, change the http_port and ssl_port to high_ports
- Change the (supplementary) group of the mympd user, myMPD should get all groups in drop privileges function

A documentation for a root-less setup for containerized MPD and myMPD would be great, also if this needs some changes in the myMPD code or myMPD docker container.

**Reply by dgcampea** [reply]
I can document it after cleaning up a bit.

One concern is that using a socket this way (monkeying with permissions) is a horrible hack, the cleanest way here seems to be just using localhost instead.

**Reply by jcorporation** [reply]
Right, changing the socket permissions in the service file is not the cleanest way. Eventually it works to set the umask to 0007 and let mpd and myMPD use the same group?

**Reply by dgcampea** [reply]
Group alone isn't enough I'm afraid, it will need to be the same user (`--userns=keep-id` might work though).

**Reply by dgcampea** [reply]
While investigating this and tweaking the reverse proxy, I think the performance problem might be due to some oversized cover images being transfered (multi-megabyte images).

Does mympd have any option to perform cover resizing on backend before returning? (caching these can be done either by nginx afterwards or on mympd through the existing 'covercache' mechanism)
I've been playing around with nginx image_filter which sort of works but has many limitations and greatly complicates the setup.
Right now I'm seeing some eye-watering 600+ MB transfered for less than 100 album covers.

**Reply by jcorporation** [reply]
myMPD has no cover resizing option. I do not like on-the-fly resizing. Better way could be to add thumbnail images to the library and myMPD choose if it delivers the thumbnail or the original image. This would involve a new configuration setting for thumbnail covernames. Would this solve your problem in a good manner?

**Reply by dgcampea** [reply]
While thinking about this, I ended up writing a small Flask shim that sits between nginx and mympd for the `/albumart` endpoint.
Right now it's at a proof of concept stage but solving this by *reverse proxy* approach allows me to:
- Force an image format for all covers (converting them to AVIF on the fly for example)
- Conditionally resize images (with imagemagick there's an ability to target by file size too)
- Conditionally crop images (using `uri=` parameter and criteria, python-pillow can be used here)
- Cache these shimmed responses with nginx
- Dynamic response through client criteria (return even smaller resized images if requested from IP range/user-agent/etc.)

I don't know if it's worth adding a custom thumbnail directory, it would need to be maintained and tracked against library changes. Not impossible but the users will have to cook up their own scripts.

**Reply by dgcampea** [reply]
There's an issue that I noticed while writing the shim, the `uri=` parameter doesn't treat `+` the same as `%20` (space). When URLdecoding and URLencoding the headers with Python (?), mympd fails to find the image even when no changes were done. This makes it tough to cleanly pass headers around in Flask (right now I'm working around it by passing the parameters as a string and crafting a url with it).

**Reply by jcorporation** [reply]
As I know the spec %20 is used for url encoding, + is for application/x-www-form-urlencoding. myMPD handle url encoding / decoding the same way as JavaScript does, using %20 for space, not +.

**Reply by dgcampea** [reply]
Still exploring the image shim idea + caching, some questions:
- Can a 'size' query parameter be added to `/albumart` URLs by the frontend that uses the values under Settings/Appearance/Albumart/{Size normal, small} depending on the request (absent if intention is requesting the full image from `Zoom` button)? The backend doesn't necessarily need to do anything about them, their purpose is to allow the reverse proxy or shim to process it and either resize or respond with an item from the cache.
- When no albumart is found, how about using code 404 instead? It allows the reverse proxy to skip caching identical data:
```
# example nginx
location /albumart {
   proxy_pass ....
   proxy_cache_key $arg_uri;
   error_page 404 /mympd_notfound.svg;
}

# or proxy_pass .../mympd_404albumart.svg; ?
location = /mympd_notfound.svg {
}
```

**Reply by jcorporation** [reply]
Generic size values would be possible thumbnail/normal/original. The shim or reverse proxy could than assign the exact dimensions.

Additional myMPD could select the right image from the filesystem, calculating the covername by appending the size to the covername, e.g. cover-sm, cover-md, cover

404 for not available cover images is a good idea if browsers displays the image anyhow.

**Reply by dgcampea** [reply]
> 404 for not available cover images is a good idea if browsers displays the image anyhow.

I tested this within the shim by replacing the original response code with 404 instead and both Chromium and Firefox handle it well, the svg cover is shown.

> Generic size values would be possible thumbnail/normal/original.

This can work too.

Did some tests, preliminary results:
Conversion on the fly slows down the initial browsing slightly from the LAN perspective (over WAN, the transfer times actually dwarf the conversion). Once converted and cached by nginx, the responses are nearly instantaneous (from ~1.5-5s before shimming to sub 50ms after shim+cache). This was with Flask+Pillow+avifenc as subprocess and uwsgi+nginx (with uwsgi_cache enabled).

Although I still need to iron out some issues with it, the shim approach should effectively "solve" the performance issue here.

Edit:
Bandwidth savings: ~100-300MB per 100 albums -> 3.7MB 

Edit 2:
Can you retest SVG responses with 404 code for firefox? I was under the impression that it was working but they don't seem to be but Chromium seems fine though.

**Reply by dgcampea** [reply]
Alternatively, if the requested cover doesn't exist, maybe respond with a 30x redirect to `/assets/coverimage-notavailable.svg`? Makes it pretty straightforward to catch them in reverse proxy.

**Reply by jcorporation** [reply]
I would prefer 404 because the image could be delivered directly, without a second connection.

**Reply by dgcampea** [reply]
Does it send a second request beyond the first unsuccessful `/albumart`?
After the first unsuccessful `/albumart` request and 30x redirect to `/assets/coverimage-notavailable.svg`, subsequent unsuccessful `/albumart` requests will be redirected to the same URL that is already cached since `/assets/coverimage-notavailable.svg` sets `cache-control: max-age=604800`.

**Reply by jcorporation** [reply]
myMPD requests the image by referencing it in a html tag or css value. The browser requests the file only once and should respect the cache-control headers.

Looking at the code: it is not simple to change the response code from `200` to `404` for the `not-available.svg`. Sending a redirect is also not so easy, because the backend does not know if it is hosted behind a reverse proxy through a subdir.

**Reply by dgcampea** [reply]
I did the following experiment:
1. Return 404 if `content-type` is `image/svg+xml` as a way to flag this particular response. (in hindsight, I could have performed the redirect here instead of playing with nginx)
2. Have nginx process 404 and redirect it elsewhere:
```
         location = /albumart {
(... proxy/uwsgi settings)
                uwsgi_intercept_errors on;
                # Doesn't work:      error_page 404 =307 /assets/coverimage-notavailable.svg
		error_page 404 = @coverimage-notavailable;
	}
	location @coverimage-notavailable {
		internal;
		return 307 /assets/coverimage-notavailable.svg;
	}
```

The network flow with this setup looks like:
1. Initial load when opening /
![load](https://user-images.githubusercontent.com/59268455/166074316-bbee2150-7856-4631-8bd5-13ce6e510cc3.png)
2. Browsing albums
![cover307](https://user-images.githubusercontent.com/59268455/166074415-243409d7-3e1c-48c3-8fdc-a661692e3d79.png)

`/assets/coverimage-notavailable.svg` is fetched only once by *sw.js* and gets reused in further requests. There's no particular reason that this must be done by the reverse proxy, I think it can be done directly in mympd instead?

--Edit
In comparison with original behavior (`svg+xml` image corresponding to `coverimage-notavailable.svg` is transferred multiple times):
![image](https://user-images.githubusercontent.com/59268455/166075968-28725f75-b1ad-40aa-a3c5-dedca4834c80.png)

This results in inefficient caching by browser (and reverse proxy if used) and bandwidth usage.

**Reply by jcorporation** [reply]
Initial fetch is from the service worker cache. Redirects to this uri could be handled directly by the serviceworker.

**Reply by jcorporation** [reply]
I am thinking about your redirect approach, ~~but I am hanging how the myMPD backend could detect the redirect path if it is served behind a reverse proxy in a subdir. Are relativ redirects working?~~

https://stackoverflow.com/questions/8250259/is-a-302-redirect-to-relative-url-valid-or-invalid

A relative redirect could be an easy solution.

**Reply by jcorporation** [reply]
Commit e91f03f adds the redirect feature.

**Reply by dgcampea** [reply]
Yes, I was using relative redirects :)

**Comment by dgcampea** [comment]
# Cover Image search

Allow the use of regex expressions and "meta variables for artists". Based on proposal https://github.com/MusicPlayerDaemon/MPD/issues/1493.

**Reply by jcorporation** [reply]
Current implementation is faster and much easier in the implementation because we must only test for file existence. With regex myMPD must read the whole folder content.

Meta variables: how they are defined?

**Reply by dgcampea** [reply]
The meta variables are based on the tag values. The idea is to search for something like `%AlbumArtist%.jpg`

**Reply by jcorporation** [reply]
I see no benefit of this feature. It adds overhead because we must first read the tag value from mpd.

**Reply by dgcampea** [reply]
The mpd library I'm using is mounted as read-only and the file naming isn't consistent, occasionally some cover files would be named in that manner.
Not too big of a problem as the shim can be adapted to do it if necessary (or maybe an overlayfs with beets/etc. to download or symlink covers appropriately, still need to think more about this).

**Reply by dgcampea** [reply]
Incidentally, is there any penalty in making the *Albumart/Name* parameter case insensitive?
Right now I'm duplicating them in the following manner:
```
folder,cover,Image,image,front,Front,Image001,image001,Image0001,image0001
```

**Reply by jcorporation** [reply]
Same penalty as for regex would apply.

**Reply by jcorporation** [reply]
In the v9.3.0 branch there is now an additional endpoint /albumart-thumb to deliver thumbnails for coverimages. Filenames for thumbnails can be configured in the settings dialog.

**Reply by dgcampea** [reply]
> In the v9.3.0 branch there is now an additional endpoint /albumart-thumb to deliver thumbnails for coverimages. Filenames for thumbnails can be configured in the settings dialog.

Thanks.

Question: What are the arguments accepted by `/albumart`? So far I only identified `uri=` as being particularly relevant but I'm not sure what is `offset=` used for.

**Reply by jcorporation** [reply]
Documentation is here: https://github.com/jcorporation/myMPD/blob/v9.3.0/docs/references/webserver-uris.md.

Set offset to `0`,  it must be the first parameter, followed by the uri param. The query string is parsed manually.

**Comment by dgcampea** [comment]
# /albumart URL normalization

This mostly concerns caching reverse proxies.

The *uri* parameter for cuesheet based tracks is of the form: `/<path-to-file>.cue/track0001`. As `/albumart` is based on the existence of `cover.jpg, ...` files under the album directory, `albumart` will always return the same image regardless of the `/track000*` value.
The (unlikely) pathological case is if a cuesheet references tracks in different directories but I'm unaware of the existence of such files. (is it even valid for cuesheets to reference files outside of their directory?)

If behind a caching reverse proxy, this ends up caching the same image if `/albumart` gets requested multiple times for different tracks of the same album.

They can be normalized within the reverse proxy (nginx snippet):
```
    if ($arg_uri ~ ^(.*)/track\d+$) {
        set $arg_uri $1;
    }

   location /albumart {
         (...)
         proxy_cache_key $arg_uri;
   }
```

Should mympd strip the `/track****` part or does it suffice to mention this in the docs (maybe under an advanced reverse-proxy section)?

**Reply by jcorporation** [reply]
Normalization is done at the backend for direct music directory access. MPD expects the full path for the albumart and readpicture commands. Normalization could break this two mpd commands.

**Comment by dgcampea** [comment]
# Mobile UX: Do not pop keyboard up in *Queue* view

Clicking on *Queue* under Firefox for Android immediately pops up the keyboard awaiting for user input for the queue search bar. Is it possible to disable this behavior? It's obstructive when switching between *Queue* and other views (such as *Playback*).

**Reply by jcorporation** [reply]
#747

**Comment by dgcampea** [comment]
# Mobile UX issue: Truncated coverart thumbnail

Thumbnail no longer fits in the bottom bar.

![icon](https://user-images.githubusercontent.com/59268455/166608814-afe66082-2857-48f1-bd67-9b81c5b23c2c.png)

**Reply by jcorporation** [reply]
#748
